File: tools/run_with_retrofit.R
#!/usr/bin/env Rscript
# =============================================================================
# 以出图改造层运行既有 R 画图脚本（路线 B 的 R 侧入口）
#
#   Rscript tools/run_with_retrofit.R <script.R> [width_in] [font_pt]
#
# 示例：
#   Rscript tools/run_with_retrofit.R 07_bulk_prognosis/07_01_prognosis_main.R 3.2 7
# =============================================================================

args <- commandArgs(trailingOnly = TRUE)

# 关键：本仓库路径含中文。R 在 Windows 下默认的 native encoding 无法转换
# 中文路径（unable to translate to native encoding / file name conversion
# problem），必须先切到 UTF-8 locale，否则读写数据文件都会失败。
invisible(Sys.setlocale("LC_CTYPE", ".UTF-8"))

if (length(args) < 1) {
  cat("用法: Rscript tools/run_with_retrofit.R <script.R> [width_in] [font_pt]\n")
  quit(status = 1)
}

# 解析 --out <目录>（位置参数顺序：script [width] [font]）
out_flag <- NULL
pos <- character(0)
i <- 1
while (i <= length(args)) {
  if (args[i] == "--out" && i < length(args)) {
    out_flag <- args[i + 1]
    i <- i + 2
  } else {
    pos <- c(pos, args[i])
    i <- i + 1
  }
}
args <- pos

script <- args[1]
if (!file.exists(script)) { cat("脚本不存在:", script, "\n"); quit(status = 1) }

# 注意：本仓库路径含中文（...\论文修改\...），R 在 Windows 下对含中文的
# **绝对路径** 做编码转换会失败（unable to translate to UTF-8）。
# 因此这里一律使用相对路径，并确保在仓库根目录运行。
RETRO_W   <- if (length(args) >= 2) as.numeric(args[2]) else 3.2
RETRO_PT  <- if (length(args) >= 3) as.numeric(args[3]) else 7
RETRO_OUT <- if (!is.null(out_flag)) out_flag else
  file.path("results", "figures",
            tools::file_path_sans_ext(basename(script)))

# 先定义变量，再 source retrofit.R（其中的函数会引用这些变量）
source(file.path("config", "retrofit.R"), local = FALSE)

cat(sprintf("[retrofit] 目标 panel 宽 %.2f in，排版后字号 %.0f pt\n", RETRO_W, RETRO_PT))
cat(sprintf("[retrofit] 输出重定向到 %s\n", RETRO_OUT))
cat(sprintf("[retrofit] 运行 %s\n", script))

source(script, local = FALSE)
cat("[retrofit] 完成\n")

File: tools/run_with_retrofit.py
"""
以出图改造层运行既有画图脚本（路线 B 的执行入口）
================================================================================
在不改动原脚本内部逻辑的前提下，把画布缩到版面尺寸并同步放大字号，
使输出的图在 LaTeX 中 1:1 显示时，文字就是设定的字号。

用法：
    python tools/run_with_retrofit.py <script.py> [target_width_in] [font_pt]

示例：
    # Fig4：按两栏布局（每个 panel 3.2 in 宽），最终字号 7 pt
    python tools/run_with_retrofit.py 05_CellOracle/05_90_reproduce_figures_local.py 3.2 7

    # 单栏大图（panel 6.3 in 宽），字号 8 pt
    python tools/run_with_retrofit.py 04_pySCENIC/04_01_pyscenic_regulons.py 6.3 8
================================================================================
"""

from __future__ import annotations

import runpy
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPO))


def main() -> None:
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)
    target = Path(sys.argv[1])
    if not target.exists():
        print(f"脚本不存在：{target}")
        sys.exit(1)

    # 解析参数：位置参数 [宽度] [字号] [输出目录]，另支持 --out <目录>
    pos, out_flag, design, plain, strip = [], None, False, False, False
    i = 1
    while i < len(sys.argv):
        if sys.argv[i] == "--out" and i + 1 < len(sys.argv):
            out_flag = sys.argv[i + 1]
            i += 2
            continue
        if sys.argv[i] == "--design":
            design = True
            i += 1
            continue
        if sys.argv[i] == "--plain":
            plain = True
            i += 1
            continue
        if sys.argv[i] == "--strip-titles":
            strip = True
            i += 1
            continue
        pos.append(sys.argv[i])
        i += 1

    # 第 1 个位置参数 = 这张图在论文中的最终显示宽度（inch），不是画布宽度。
    # 画布保持脚本原尺寸，字号与图形元素按此反推同步放大。
    width = float(pos[1]) if len(pos) > 1 else 6.3
    font = float(pos[2]) if len(pos) > 2 else 7.0

    # 输出目录：默认 results/figures/<脚本名>，可覆盖（--out 优先）。
    # 目的是绝不覆盖用户已有的原始图片。
    out = (out_flag
           if out_flag
           else (pos[3] if len(pos) > 3
                 else str(REPO / "results" / "figures" / target.stem)))

    import config.retrofit as retrofit
    retrofit.install(target_display_in=width, target_font_pt=font,
                     output_dir=out, design=design, plain=plain,
                     strip_titles=strip)

    print(f"[retrofit] 最终显示宽 {width} in，排版后字号 {font} pt")
    print("[retrofit] 模式：" + ("原样输出（保持脚本原始尺寸与字号）" if plain else
                                ("定尺设计（画布=最终显示尺寸）" if design else "缩画布保字号")))
    print(f"[retrofit] 输出重定向到 {out}")
    print(f"[retrofit] 运行 {target}")
    runpy.run_path(str(target), run_name="__main__")


if __name__ == "__main__":
    main()
